### ISSUE TITLE
requiring "dd-trace/ci/jest/env" causes Jest to exit with non-zero exit status

### ISSUE THREAD

**Issue by gurchik** [DataDog/dd-trace-js#1854]
**Describe the bug**
I'm following the [documentation here](https://docs.datadoghq.com/continuous_integration/setup_tests/javascript/?tab=jest) to instrument my Jest tests. I've added `testEnvironment: '<rootDir>/testEnvironment.js',` to my jest config, and my testEnvironment.js is as follows (from the documentation):

```js
if (process.env.DD_ENV === 'ci') {
  //require('dd-trace/ci/jest/env');
}

module.exports = require('jest-environment-jsdom');
```

My tests pass just fine. But if I uncomment the line, the tests still all pass, but Jest exits with a non-zero exit status, causing my CI to fail.

```
Test Suites: 90 passed, 90 total
Tests:       637 passed, 637 total
Snapshots:   0 total
Time:        442.004 s
Ran all test suites.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: 1

Exited with code exit status 1
CircleCI received exit code 1
```

Despite the failure, the traces do appear to be working:
![Screen Shot 2022-02-17 at 1 49 28 PM](https://user-images.githubusercontent.com/602444/154550459-1a842936-9c22-40ed-b391-3c59418c363b.png)

I've uncommented and recommented this line several times and it has a 100% reproduction rate for me.

**Environment**

* **Operation system:** Alpine 3.15 (specifically, the docker image `node:16.13-alpine3.15`
* **Node version:** 16.13
* **Tracer version:** 2.2.0
* **Agent version:** the `gcr.io/datadoghq/agent:latest` image (shouldn't be relevant though)

Additionally, here are the relevant bits from my yarn.lock:

```
dd-trace@^2.2.0:
  version "2.2.0"
  resolved "https://registry.yarnpkg.com/dd-trace/-/dd-trace-2.2.0.tgz#414b02a05c2ea8f6a9304f267d92834fb00cc7ce"
  integrity sha512-PEYucl6bQoyz4CF6kCu1LBbJ9p8qec7sl6cbnfV+wZoMgdQrf30q/iV+5YbyzVCZVnkY5xsFqR8f08/V/ygIBw==

jest@^27.4.5:
  version "27.4.7"
  resolved "https://registry.yarnpkg.com/jest/-/jest-27.4.7.tgz#87f74b9026a1592f2da05b4d258e57505f28eca4"
  integrity sha512-8heYvsx7nV/m8m24Vk26Y87g73Ba6ueUd0MWed/NXMhSZIm62U/llVbS0PJe1SHunbyXjJ/BqG1z9bFjGUIvTg==

jest-circus@^27.4.6:
  version "27.4.6"
  resolved "https://registry.yarnpkg.com/jest-circus/-/jest-circus-27.4.6.tgz#d3af34c0eb742a967b1919fbb351430727bcea6c"
  integrity sha512-UA7AI5HZrW4wRM72Ro80uRR2Fg+7nR0GESbSI/2M+ambbzVuA63mn5T1p3Z/wlhntzGpIG1xx78GP2YIkf6PhQ==

jest-environment-jsdom@^27.4.6:
  version "27.4.6"
  resolved "https://registry.yarnpkg.com/jest-environment-jsdom/-/jest-environment-jsdom-27.4.6.tgz#c23a394eb445b33621dfae9c09e4c8021dea7b36"
  integrity sha512-o3dx5p/kHPbUlRvSNjypEcEtgs6LmvESMzgRFQE6c+Prwl2JLA4RZ7qAnxc5VM8kutsGRTB15jXeeSbJsKN9iA==
```

I tried adding DD_TRACE_DEBUG=true but didn't see any obvious errors.

**Comment by rochdev** [comment]
@juan-fernandez Do we do anything that could cause an error on process exit or after the test suite is done?

**Comment by juan-fernandez** [comment]
hey @gurchik! Thanks for reporting this issue. I haven't been able to reproduce locally. What we can try next: 

1. In addition to `DD_TRACE_DEBUG=true`, could you add `DD_TRACE_LOG_LEVEL=error` to your env vars? We might see something interesting in the logs. 
2. Could you share your `jest` configuration? Feel free to remove any field you think should be private. 
3. Can you reproduce this issue with a smaller subset of tests? 

Only thing I can think of is the `flush` is throwing an error when the test suite is done. I've pushed a branch to test this: https://github.com/DataDog/dd-trace-js/compare/juan-fernandez/attempt-fix-error-code. Since I can't reproduce this, would you mind testing this branch in your CI? You can install it via `npm install https://github.com/DataDog/dd-trace-js/tarball/juan-fernandez/attempt-fix-error-code`

**Comment by gurchik** [comment]
Hi @juan-fernandez thanks for looking into this.

I tried with `DD_TRACE_LOG_LEVEL=error` but all that did is hide the debug statements from before. Looks like the default is `debug` so those are not printed with `error`. Still no errors though.

Here is my jest config:

```js
/* eslint-disable max-len */
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

const path = require('path');

module.exports = {
  clearMocks: true,

  collectCoverageFrom: [
    'src/**/*.{js,jsx,mjs}',
    '!src/**/*styles.{js,jsx,mjs}',
    '!src/**/*Styles.{js,jsx,mjs}',
  ],
  coverageDirectory: 'coverage',
  coveragePathIgnorePatterns: [
    '\\\\node_modules\\\\',
    'src/REDACTED/',
  ],
  coverageReporters: ['json', 'lcov'],

  globals: {
    REDACTED: "redacted",
  },

  moduleDirectories: [
    'node_modules',
    path.resolve(__dirname, 'src'),
  ],
  moduleFileExtensions: ['js', 'json', 'jsx'],
  moduleNameMapper: {
    d3: '<rootDir>/node_modules/d3/dist/d3.min.js',
  },
  setupFiles: ['<rootDir>/enzyme.config.js'],
  setupFilesAfterEnv: ['<rootDir>/jestSetup.js'],

  testEnvironment: '<rootDir>/.testEnvironment.js',
  //testEnvironment: 'jsdom',

  testMatch: ['**/__tests__/**/*.js?(x)', '**/?(*.)+(spec|test).js?(x)'],
  testPathIgnorePatterns: ['\\\\node_modules\\\\', 'src/REDACTED/'],
  testURL: 'http://localhost',
  transform: {
    '^.+\\.js$': 'babel-jest',
    '.+\\.(css|styl|less|sass|scss)$': 'jest-transform-css',
  },
  transformIgnorePatterns: ['<rootDir>/node_modules/*.css'],
  verbose: false,
};
```

I can try your branch and will let you know. I wouldn't expect `flush` to cause a problem because the test results _are_ flushed successfully to DD, but you never know!

**Comment by josh--newman** [comment]
Hey there @juan-fernandez and @gurchik. I've been having the same issue and tried to test with the test branch linked above but haven't seen any improvement. Here's my `jest.config.ts` file:

```
import type { Config } from '@jest/types';

const config: Config.InitialOptions = {
  collectCoverageFrom: [
    'src/**/*.{js,ts,tsx}',
    '!src/**/*.stories.{js,ts,tsx}',
  ],
  coverageReporters: ['html', 'text', 'cobertura'],
  globals: {
    'ts-jest': {
      babelConfig: true,
      isolatedModules: true,
    },
  },
  rootDir: '.',
  roots: ['<rootDir>/src/'],
  preset: 'ts-jest',
  setupFilesAfterEnv: [
    '@testing-library/jest-dom/extend-expect', // Add some helpful assertions. See: https://testing-library.com/docs/ecosystem-jest-dom
    '<rootDir>/jest/setup.ts',
  ],
  testEnvironment: '<rootDir>/jest/testEnvironment.js',
  testTimeout: 15_000,
};

export default config;

```

**Comment by josh--newman** [comment]
I solved this problem for my specific case. I had a complicated React component test that was unintentionally making a network request asynchronously upon rendering. All the tests passed but the node process ultimately failed because the request promise threw an error _after_ the tests finished running.

Debugging was further complicated by the fact that the actual error stack trace was being masked due to some hackery on my part.

**Comment by juan-fernandez** [comment]
hey @gurchik! I'm still struggling to reproduce. I can get tests to pass and still have an error code (https://github.com/facebook/jest/issues/9324), but I can't make it depend on the initialisation of `dd-trace`. I have some questions:

* Does this happen only in CI or also locally?
* Is there any way for you to reproduce with a smaller set of tests? If we can identify what those tests are doing I might be able to reproduce. 
* Running tests with `--detectOpenHandles` (https://jestjs.io/docs/cli#--detectopenhandles) might be helpful to identify process exit issues. Could you run your suite with that?

**Comment by gurchik** [comment]
Hey @juan-fernandez thanks for continuing to look into this. I'll give those a try and get back soon.

**Comment by juan-fernandez** [comment]
maybe related: https://github.com/DataDog/dd-trace-js/issues/1945

**Comment by juan-fernandez** [comment]
hey @gurchik, could you try out [2.4.2](https://github.com/DataDog/dd-trace-js/releases/tag/v2.4.2)? There has been a change in the way the jest environment is tore down and it might fix your issue.

**Comment by gurchik** [comment]
@juan-fernandez yes that appears to have fixed it! I tried on 2.2.0 and confirmed it was still not working properly, then I upgraded to 2.4.2 and it works great. Seems to be consistently working. Thank you!

**Comment by juan-fernandez** [comment]
that's awesome news! 🎉 

I'll close this issue then. Feel free to comment it here if the issue reappears
